src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts

```diff
@@ -2,244 +2,259 @@ import * as React from 'react';
+     * @param filter Text used to filter suggestions.
+     */
+    private searchTags(filter?: string): Promise<ITag[]> {
+        let options = `?$select=${this.idAttribute},${this.nameAttribute}&$orderby=${this.nameAttribute} asc`;
+
+        if (filter)
+            options = `${options}&$filter=contains(${this.nameAttribute},'${filter}')`;
+
+        return this.webAPI.retrieveMultipleRecords(this.relatedEntity, options).then(
+            results => {
+                if (results.entities.length < 1)
+                    return [];
+
+                return results.entities.map(item => ({ key: item[this.idAttribute], name: item[this.nameAttribute] }));
+            }
+        );
+    }
+
+    /**
+     * A callback for when the selected list of items changes.
+     * @param items A collection containing the items.
+     */
+    private onChange(items?: ITag[]) : void {
+        const promises: Promise<Response>[] = [];
+        const entityExists: boolean = (this.entityId !== undefined && this.entityId !== "00000000-0000-0000-0000-000000000000");
+
+        // We only need to associate / dissasociate items when the entity exists.
+        if (entityExists)
+        {
+            // Associate the added items.
+            const itemsAdded = items?.filter(item => !this.selectedItems.some(selectedItem => selectedItem.key === item.key)) || [];
+            for(let item of itemsAdded) {
+                promises.push(this.associateItem(item));
+            }
+
+            // Disassociate the removed items.
+            const itemsRemoved = this.selectedItems.filter(selectedItem => !items?.some(item => item.key === selectedItem.key));
+            for (let item of itemsRemoved) {
+                promises.push(this.disassociateItem(item));
+            }
+        }
+
+        Promise.all(promises).then(
+            results => {
+                this.props.selectedItems = this.selectedItems = items || [];
+
+                if (!entityExists)
+                    this.notifyOutputChanged();
+            }
+        );
+    }
+
+    /**
+     * Associate the item with the entity.
+     * @param item The item to associate.
+     */
+    private associateItem(item: ITag): Promise<Response> {
+        const parentSetName: string = this.entityMetadata[EntityMetadataProperties.EntitySetName];
+        const childSetName: string = this.relatedEntityMetadata[EntityMetadataProperties.EntitySetName];
+
+        return this.webAPI.associateRecord(parentSetName, this.entityId, this.relationshipName, childSetName, item.key);
+    }
+
+    /**
+     * Disassociate the item with the entity.
+     * @param item The item to disassociate.
+     */
+    private disassociateItem(item: ITag): Promise<Response> {
+        const parentSetName: string = this.entityMetadata[EntityMetadataProperties.EntitySetName];
+
+        return this.webAPI.disassociateRecord(parentSetName, this.entityId, this.relationshipName, item.key);
+    }
```

Comment:
now that the code is so simple, we can probably move that back to the for loop.

Reply:
The methods have been moved inside the loops